docs/design/design_v1.8.md

```diff
@@ -0,0 +1,437 @@
+- Remove the Static Pod manifest file. The kubelet will stop the original Static Pod-hosted component that was running
+
+Please note that:
+
+1. Self hosting is not yet resilient to node restarts; this can be fixed with external checkpointing; in 1.9 kubelet checkpointing will be available as well
+2. If invoked with `—features-gates=StoreCertsInSecrets`  following additional steps will be executed
+   - creation of `ca`,  `apiserver`,  `apiserver-kubelet-client`, `sa`, `front-proxy-ca`, `front-proxy-client`  TLS secrets in `kube-system` namespace with respective certificates and keys
+   - creation of `schedler.conf` and `controller-manager.conf` secrets in`kube-system` namespace with respective kubeconfig files
+   - mutation of  all the pod specs by replacing host path volumes with projected volumes from the secrets above
+3. This phase can be invoked individually with the `kubeadm phase selfhosting convert-from-staticpods`  command.
+4. In case of `kubeadm`  executed in the `--dry-run` mode, this phases targets a "fake API server".
+
+## `kubeadm join` phases
+
+Similarly to `kubeadm init`, also `kubeadm join` internal workflow consists of a sequence of atomic work tasks to perform.
+
+This is split into discovery (having the Node trust the Kubernetes Master) and TLS bootstrap (having the Kubernetes Master trust the Node).
+
+### Discovery cluster-info
+
+There are 2 main schemes for discovery. The first is to use a shared token along with the IP address of the API server. The second is to provide a file (a subset of the standard kubeconfig file). 
+
+#### Shared token discovery
+
+If `kubeadm join` is invoked with `--discovery-token`, token discovery is used; in this case the node basically retrieves the cluster CA certificates from the  `cluster-info` ConfigMap in the `kube-public` namespace.
+
+In order to prevent "man in the middle" attacks, several steps are taken:
+
+- First, the CA certificate is retrived via insecure connection (NB. this is possible because `kebeadm init` granted access to  `cluster-info` users for `system:unauthenticated` )
+- Then the CA certificate goes trough following validation steps: 
+  - "Basic validation", using the token ID against a JWT signature
+  - "Pub key validation", using provided `--discovery-token-ca-cert-hash`. This value is available in the output of "kubeadm init" or can be calcuated using standard tools (the hash is calculated over the bytes of the Subject Public Key Info (SPKI) object as in RFC7469). The `--discovery-token-ca-cert-hash flag`may be repeated multiple times to allow more than one public key.
+  - as a additional validation, the CA certificate is retrived via secure connection and then compared with the CA retrieved initially
+
+Please note that:
+
+1.  "Pub key validation" can be skipped passing `--discovery-token-unsafe-skip-ca-verification` flag; This weakens the kubeadm security model since others can potentially impersonate the Kubernetes Master.
```

Comment:
This file is then written down to `/etc/kubernetes/bootstrap-kubelet.conf` that the kubelet uses to do TLS Bootstrapping